File: src/CCVTAC.FSharp/Downloading/Updater.fs
namespace CCVTAC.Console.Downloading

open CCVTAC.Console.ExternalTools
open CCVTAC.Console
open CCVTAC.Console.Settings.Settings

module Updater =

    let run userSettings (printer: Printer) : Result<unit,string> =
        if String.hasNoText userSettings.DownloaderUpdateCommand then
            printer.Info("No downloader update command provided, so will skip.")
            Ok()
        else
            let toolSettings = ToolSettings.create userSettings.DownloaderUpdateCommand userSettings.WorkingDirectory

            match Runner.run toolSettings [] printer with
            | Ok result ->
                if result.ExitCode <> 0 then
                    printer.Warning("Tool updated with minor issues.")

                    match result.Error with
                    | Some w -> printer.Warning w
                    | None -> ()

                Ok()

            | Error err ->
                printer.Error $"Failure updating: {err}"
                Error err

File: src/CCVTAC.Console/Downloading/Updater.cs
using CCVTAC.Console.ExternalTools;
using UserSettings = CCVTAC.FSharp.Settings.UserSettings;

namespace CCVTAC.Console.Downloading;

internal static class Updater
{
    private record Urls(string Primary, string? Supplementary);

    /// <summary>
    /// Completes the actual download process.
    /// </summary>
    /// <returns>A `Result` that, if successful, contains the name of the successfully downloaded format.</returns>
    internal static Result<string?> Run(UserSettings settings, Printer printer)
    {
        if (string.IsNullOrWhiteSpace(settings.DownloaderUpdateCommand))
        {
            printer.Info("No downloader update command provided, so will skip.");
            return Result.Ok();
        }

        var args = new ToolSettings(settings.DownloaderUpdateCommand, settings.WorkingDirectory!);

        var result = Runner.Run(args, otherSuccessExitCodes: [], printer);

        if (result.IsSuccess)
        {
            var (exitCode, warnings) = result.Value;

            if (exitCode != 0)
            {
                printer.Warning("Update completed with minor issues.");

                if (warnings.HasText())
                {
                    printer.Warning(warnings);
                }
            }

            return Result.Ok();
        }

        printer.Error($"Failure updating...");

        var errors = result.Errors.Select(e => e.Message).ToList();

        if (errors.Count != 0)
        {
            result.Errors.ToList().ForEach(e => printer.Error(e.Message));
        }

        return errors.Count > 0
            ? Result.Fail(string.Join(" / ", errors))
            : Result.Ok();
    }
}
